Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Math.NumberTheory.Primes.Factorisation.LinearAlgebra
+  ( SBVector(..)
+  , DBVector(..)
+  , SBMatrix(..)
+  , dot
+  , mult
+  , linearSolve
+  ) where
+
+#if __GLASGOW_HASKELL__ < 803
+import Data.Semigroup
+#endif
+import qualified Data.List as L
+import qualified Data.Vector.Sized as SV
+import qualified Data.Vector.Unboxed as U
+import qualified Data.Vector.Unboxed.Sized as SU
+import qualified Data.Vector.Unboxed.Mutable as MU
+import qualified Data.Vector.Unboxed.Mutable.Sized as SMU
+import Math.NumberTheory.Utils.FromIntegral
+import Control.Monad.ST
+import GHC.TypeNats (Nat, KnownNat, natVal)
+import Data.Proxy
+import System.Random
+import Data.Foldable
+import Data.Mod.Word
+import Unsafe.Coerce
+import Data.Maybe
+import Data.Bit
+import Data.Bits
+
+-- | Sparse Binary Vector of size @k.
+newtype SBVector (k :: Nat) = SBVector { unSBVector :: U.Vector (Mod k) }
+
+-- | Dense Binary Vector of size @k@.
+newtype DBVector (k :: Nat) = DBVector { unDBVector :: SU.Vector k Bit }
+  deriving (Eq, Show)
+
+-- | Sparse Binary square Matrix of size @k@. It is formed of columns
+-- of sparse binary vectors.
+newtype SBMatrix (k :: Nat) = SBMatrix { unSBMatrix :: SV.Vector k (SBVector k) }
+
+-- | Addition of two dense Binary Vectors.
+instance KnownNat k => Semigroup (DBVector k) where
+  DBVector v1 <> DBVector v2 = DBVector $ SU.withVectorUnsafe (zipBits xor (SU.fromSized v1)) v2
+
+-- | Dense Binary Vectors of given length form a group under addition.
+instance KnownNat k => Monoid (DBVector k) where
+  mempty = DBVector $ SU.replicate (Bit False)
+  mappend = (<>)
+
+-- | Dot product of two dense Binary Vectors of the same size.
+dot :: KnownNat k => DBVector k -> DBVector k -> Bit
+dot (DBVector v1) (DBVector v2) = Bit $ odd . countBits $ zipBits (.&.) (SU.fromSized v1) (SU.fromSized v2)
+
+-- | Multiplication of a square matrix and a dense vector.
+mult :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k
+mult matrix vector = runST $ do
+  vs <- SMU.new
+  traverse_ (U.mapM_ (flipBit' vs) . unSBVector . (matrix `index'`)) $ listBits' vector
+  ws <- SU.unsafeFreeze vs
+  pure $ DBVector ws
+
+listBits' :: KnownNat k => DBVector k -> [Mod k]
+listBits' = unsafeCoerce listBits
+
+flipBit' :: KnownNat k => SMU.MVector k s Bit -> Mod k -> ST s ()
+flipBit' = unsafeCoerce (unsafeFlipBit :: MU.MVector s Bit -> Int -> ST s ())
+
+index' :: KnownNat k => SBMatrix k -> Mod k -> SBVector k
+index' = unsafeCoerce SV.index
+
+-- | It takes a random seed and a square singular matrix and it returns an
+-- elemnent of its kernel. It does not check if the matrix is singular.
+linearSolve :: KnownNat k => StdGen -> SBMatrix k -> DBVector k
```

Comment:
Let's separate concerns. The caller of `linearSolve` does not neccessarily know which method of random generation is employed by `LinearAlgebra` and really should not care. Please refactor `linearSolve` to consume `seed :: Int` and construct `StdGen` internally.